48 Экран чужого проекта

TanyaRul opened the pull request.

TanyaRul committed:
- adds screen layout for a project
- adds DI files

TanyaRul committed:
- adds a repository and interactor for the view screen of any project
- renames files from Project to AnyProject
- adds a transition from the projects screen to viewing any project
- changes the layout of the viewing screen of any project

TanyaRul committed:
Merge branch 'develop' into 48-project-view-screen

# Conflicts:
#	app/src/main/res/navigation/app_nav_graph.xml

TanyaRul committed:
- adds mock data for the project on the NewsDetails screen
- adds transitions from the NewsDetails screen to the AnyProject screen and back
- removes the transition from the ProjectsScreen to the AnyProject screen

Secur1ty-21 commented on app/src/main/res/layout/fragment_any_project.xml:
```diff
@@ -0,0 +1,152 @@
+                android:layout_height="wrap_content"
+                android:layout_marginVertical="@dimen/margin_8"
+                android:background="@drawable/bg_project_card_white"
+                android:padding="@dimen/padding_16"
```
Не соответствует дизайну, у title отступ от верха карточки 12dp

TanyaRul replied:
Исправила

Secur1ty-21 commented on app/src/main/res/layout/fragment_any_project.xml:
```diff
@@ -0,0 +1,152 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Можно заменить на более легковесный LinearLayout

TanyaRul replied:
Изменила

Secur1ty-21 commented on app/src/main/res/layout/fragment_any_project.xml:
```diff
@@ -0,0 +1,152 @@
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:orientation="vertical"
+                app:layout_constraintGuide_end="@dimen/margin_8" />
```
Можно заменить на LinearLayout + paddingHorizontal="8dp"

TanyaRul replied:
Изменила

Secur1ty-21 commented on app/src/main/res/layout/fragment_any_project.xml:
```diff
@@ -0,0 +1,152 @@
+                android:layout_width="@dimen/size_0"
+                android:layout_height="wrap_content"
+                android:layout_marginVertical="@dimen/margin_8"
+                android:visibility="visible"
```
По умолчанию вроде visible

TanyaRul replied:
Исправила

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/feed/projectScreen/presentation/AnyProjectViewModel.kt:
```diff
@@ -0,0 +1,37 @@
+                    }
+                }
+            }
+        }
```
Думаю стоит добавить Dispatchers.IO и обернуть вызов useCase в runCatching
```kotlin
viewModelScope.launch(Dispatchers.IO) {
            runCatching {
                   // Вызов useCase
            }.onFailure { exception ->
                  // Обработка исключений, если вдруг такие возникли
            }
}
```

TanyaRul replied:
Добавила

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/feed/projectScreen/presentation/AnyProjectScreenState.kt:
```diff
@@ -0,0 +1,12 @@
+
+    data class Error(val type: ErrorType) : AnyProjectScreenState
+
+    data class Content(val anyProject: AnyProjectModel) : AnyProjectScreenState
```
Думаю именно data class тут необязательно, можно просто class

TanyaRul replied:
Изменила

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/feed/projectScreen/presentation/AnyProjectFragment.kt:
```diff
@@ -0,0 +1,167 @@
+
+    override val viewModel by injectViewModel<AnyProjectViewModel>()
+
+    private var newsAdapter: NewsAdapter? = null
```
Думаю лучше инициализировать сразу, чтобы не был nullable

TanyaRul replied:
Изменила

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/feed/projectScreen/presentation/AnyProjectFragment.kt:
Думаю для ленты публикаций тоже лучше сразу сделать репозитория и usecase для Мока

TanyaRul replied:
Добавила

Secur1ty-21 commented on app/src/main/java/com/innoprog/android/feature/feed/projectScreen/domain/AnyProjectModel.kt:
Над каждым полем лучше добавить аннотацию, если сделать обфускацию через proguard, то имена полей будут изменены на рандомные и gson не сможет раскрасить ответ и положить в этот класс.
```kotlin
 @SerializedName("id")
val id: String,
 @SerializedName("name")
val name: String,
```

TanyaRul replied:
Поняла. А есть ещё вариант использовать Parcelable.  Что лучше?

Secur1ty-21 replied:
Варианта Parcelable нету, он для сериализации и десериализации, ты наверное перепутала SerializedName с Serializable.

Вот что делает SerializedName 
https://stackoverflow.com/questions/28957285/what-is-the-basic-purpose-of-serializedname-annotation-in-android-using-gson

TanyaRul replied:
Упс, точно. Добавила

Secur1ty-21 reviewed (changes requested):
Если замокать 100 публикаций и перейти на этот экран то приложение виснет (ANR), необходимо оптимизировать верстку. Параметры моего устройства: Android 14, 8 + 256, Snapdragon 695.

TanyaRul committed:
- adds @SerializedName annotation to AnyProjectModel
- changes the initialization of the NewsAdapter in AnyProjectFragment, FeedFragment
- adds a repository, interactor and their implementation to the News Feed
- changes AnyProject layout
- changes the OnClickListener function on the news in NewsAdapter and NewsViewHolder
- adds Dispatchers.IO and wraps the getAnyProject function call in runCatching in AnyProjectViewModel
